File: Back/Apis/Api_Py/src/models/tablas_sql/tabla_usuarios.py
#Librerias Instaladas
from sqlalchemy import ForeignKey, String;
from sqlalchemy.orm import Mapped, mapped_column, relationship;

#Tablas Creadas
from src.models.tablas_sql.tabla_base import Tabla_Base;

#Modulos Nativos
from typing import List, Optional;
from datetime import datetime;

class Tabla_Usuarios(Tabla_Base):
    __tablename__ = 'usuarios';

    usuario_id: Mapped[int] = mapped_column(nullable=False, primary_key=True, autoincrement=True);
    usuario_nombre: Mapped[str] = mapped_column(String(50),nullable=False, unique=True);
    usuario_ultima_conexion: Mapped[datetime] = mapped_column(nullable=False);
    rol_id_fk: Mapped[int] = mapped_column(ForeignKey('roles.rol_id'), nullable=False, default=3);

    restriccion_roles_en_usuarios: Mapped['Tabla_Roles'] = relationship(back_populates='restriccion_roles_para_usuarios');

    restriccion_usuarios_para_tokens: Mapped[List['Tabla_Tokens']] = relationship(back_populates='restriccion_usuarios_en_tokens');
    restriccion_usuarios_para_archivos: Mapped[List['Tabla_Archivos_Firmados']] = relationship(back_populates='restriccion_usuarios_en_archivos');
    restriccion_usuarios_para_firmas: Mapped[List['Tabla_Firmas']] = relationship(back_populates='restriccion_usuarios_en_archivos');
    restriccion_usuarios_para_sesion: Mapped[List['Tabla_Inicio_Sesion']] = relationship(back_populates='restriccion_usuarios_en_sesion');
    restriccion_usuarios_para_acciones_sistema: Mapped[List['Tabla_Acciones_Sistema']] = relationship(back_populates='restriccion_usuarios_en_acciones_sistema');

    def __repr__(self) -> str:
        return f"usuario_id:{self.usuario_id!r}, usuario_nombre:{self.usuario_nombre!r}, usuario_ultima_conexion:{self.usuario_ultima_conexion!r}, rol_id_fk:{self.rol_id_fk!r}"

File: Back/Apis/Api_Js/src/modules/baseDaatos/tablas/tabla_usuarios.js
//Configuraciones Creadas
import { conexionBaseDatos } from "../../../settings/consultas_sql/conexion_sql.js";

//Librerias
import { DataTypes } from "sequelize";

//Modelos Creados
//--tablas sql
import { tabla_correos_usuarios } from "./tabla_correos_usuarios.js";
import { tabla_tokens } from "./tabla_tokens.js";
import { tabla_acciones_sistema } from "./tabla_acciones_sistema.js";
import { tabla_archivos_firmados } from "./tabla_archivos_firmados.js";
import { tabla_inicio_sesion } from "./tabla_inicio_sesion.js";
import { tabla_firmas } from "./tabla_firmas.js";

export const tabla_usuarios = conexionBaseDatos.define(
    'usuarios',
    {
        usuario_id:{
            type: DataTypes.INTEGER,
            primaryKey: true, 
            allowNull: false,
            autoIncrement: true,
        },
        usuario_nombre:{
            type: DataTypes.STRING,
            allowNull: false, 
            unique: true, 
        },
        usuario_ultima_conexion:{
            type:DataTypes.DATE,
            allowNull:false
        },
        rol_id_fk:{
            type:DataTypes.INTEGER,
            allowNull:false
        }
    },
    {
        freezeTableName: true,
        timestamps: false,
    }
)

tabla_usuarios.hasMany(tabla_correos_usuarios, {
    foreignKey:"usuario_id_fk"
})

tabla_usuarios.hasMany(tabla_tokens, {
    foreignKey:"usuario_id_fk"
})

tabla_usuarios.hasMany(tabla_acciones_sistema, {
    foreignKey:"usuario_id_fk"
})

tabla_usuarios.hasMany(tabla_archivos_firmados, {
    foreignKey:"usuario_id_fk"
})

tabla_usuarios.hasMany(tabla_inicio_sesion, {
    foreignKey:"usuario_id_fk"
})

tabla_usuarios.hasOne(tabla_firmas, {
    foreignKey:"usuario_id_fk"
})

tabla_correos_usuarios.belongsTo(tabla_usuarios); 
tabla_tokens.belongsTo(tabla_usuarios); 
tabla_acciones_sistema.belongsTo(tabla_usuarios);
tabla_archivos_firmados.belongsTo(tabla_usuarios); 
tabla_inicio_sesion.belongsTo(tabla_usuarios); 
tabla_firmas.belongsTo(tabla_usuarios);